File: src/content/algorithms/knapsack/code.py
def knapsack(weights, values, capacity):  # #@function-def
    n = len(weights)
    dp = [[0] * (capacity + 1) for _ in range(n + 1)]  # #@init-dp

    for i in range(1, n + 1):  # #@outer-loop
        for w in range(capacity + 1):  # #@inner-loop
            if weights[i - 1] > w:  # #@weight-check
                dp[i][w] = dp[i - 1][w]
            else:
                skip = dp[i - 1][w]
                take = dp[i - 1][w - weights[i - 1]] + values[i - 1]
                dp[i][w] = max(skip, take)  # #@decision

    # Backtrack para reconstruir os itens
    chosen = []  # #@backtrack
    i, w = n, capacity
    while i > 0:
        if dp[i][w] != dp[i - 1][w]:
            chosen.append(i - 1)
            w -= weights[i - 1]
        i -= 1

    return dp[n][capacity], list(reversed(chosen))

File: src/content/algorithms/knapsack/code.rs
pub fn knapsack(weights: &[i32], values: &[i32], capacity: i32) -> i32 {  // #@generic-function
    let n = weights.len();
    let w_cap = capacity as usize;
    let mut dp = vec![vec![0i32; w_cap + 1]; n + 1];  // #@init-dp

    for i in 1..=n {  // #@outer-loop
        let wi = weights[i - 1] as usize;
        let vi = values[i - 1];
        for w in 0..=w_cap {  // #@inner-loop
            if wi > w {  // #@weight-check
                dp[i][w] = dp[i - 1][w];
            } else {
                let skip = dp[i - 1][w];
                let take = dp[i - 1][w - wi] + vi;
                dp[i][w] = skip.max(take);  // #@decision
            }
        }
    }

    // Backtrack
    let mut w = w_cap;  // #@backtrack
    let mut i = n;
    while i > 0 {
        if dp[i][w] != dp[i - 1][w] {
            w -= weights[i - 1] as usize;
        }
        i -= 1;
    }

    dp[n][w_cap]
}
